### PR TITLE
fix: do not wrap invalid grammar objects during constrained generation

### PR DESCRIPTION
**Pull request by tazjin**
Wrapping invalid grammar objects in the reasoner backend causes the scheduler's check for invalid grammars to not notice them.

This can lead to requests with invalid grammars (e.g. on regex validity issues) to be scheduled, eventually causing a crash when the invalid grammar is used for masking.

This relates to sgl-project/sglang#9187

## Motivation

This pull request is motivated by a production incident in which a user managed to bring down several LLM instances (DeepSeek-3.2 and GLM-4.6) by supplying the following JSON schema in a request:

```
{"type": "object", "properties": {"reasoning": {"type": "string", "minLength": 16, "description": "A SINGLE multiline plain-text string. Use \'- \' for bullet points inside the string. Do NOT return arrays or objects here.", "pattern": "^(?!\\\\s*[\\\\[{]).+"}, "test_code_lines": {"type": "array", "items": {"type": "string"}, "minItems": 1, "description": "Full test code, split line-by-line. Each array element is one code line."}}, "required": ["reasoning", "test_code_lines"], "additionalProperties": false}
```

The regular expression here causes `/project/cpp/regex_converter.cc:75: Regex parsing error at position 4: Lookahead is not supported yet.`, produces an invalid grammar object, and later sglang attempts to call methods on this object.

## Modifications

Ensured that we don't wrap invalid grammar objects in the reasoner grammar wrapper. This way further checks for invalid grammar objects "further down the pipe" will work as intended.

## Accuracy Tests

We tried sending requests with this JSON schema after the patch, and no more crashes were observed.

### CODE REVIEW

File: python/sglang/srt/constrained/reasoner_grammar_backend.py [hunk 1]
```diff
@@ -85,6 +89,7 @@ def _init_value_dispatch(
         self, key: Tuple[str, str]
     ) -> Optional[ReasonerGrammarObject]:
         ret = self.grammar_backend._init_value_dispatch(key)
-        if ret is None:
-            return None
+        # avoid wrapping invalid grammar, so that the scheduler can detect it
+        if ret is None or ret is INVALID_GRAMMAR_OBJ:
```
[the remark this answers is not in the document]
Comment (tazjin) [reply python/sglang/srt/constrained/reasoner_grammar_backend.py:91]: Happy to do this if a test confirms it / or a human with Python knowledge confirms it. I don't normally use dynamically typed languages, so not sure what side effects this might have.

[the remark this answers is not in the document]
Comment (JustinTong0323) [reply python/sglang/srt/constrained/reasoner_grammar_backend.py:91]: make sense actually

[the remark this answers is not in the document]
Comment (tazjin) [reply python/sglang/srt/constrained/reasoner_grammar_backend.py:91]: Done!

File: python/sglang/srt/constrained/reasoner_grammar_backend.py [hunk 2]
```diff
@@ -17,7 +17,11 @@
 
 import torch
 
-from .base_grammar_backend import BaseGrammarBackend, BaseGrammarObject
+from .base_grammar_backend import (
```
Comment (hnyls2002) [inline python/sglang/srt/constrained/reasoner_grammar_backend.py:20]: Also, change the import to `from sglang.srt.constrained.base_grammar_backend import xxx`

**Review by hnyls2002 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/constrained/reasoner_grammar_backend.py (hunk 2 above)
Comment (tazjin) [reply python/sglang/srt/constrained/reasoner_grammar_backend.py:20]: I won't be able to test that everything works with another change until next week. The import was already like this before my PR, but it's up to you whether to block the fix on this or not.
